lib/Db/TagMapper.php

```diff
@@ -166,6 +164,40 @@ function (string $key) {
 		return $tags;
 	}
 
+	/**
+	 * @param Message[] $messages
+	 * @param string $userId
+	 * @param string $imapLabel
+	 * @return string[]
+	 */
+	public function getTaggedMessageIdsForMessages(array $messages, string $userId, string $imapLabel): array {
+		$ids = array_map(static function (Message $message) {
+			return $message->getMessageId();
+		}, $messages);
+
+		$qb = $this->db->getQueryBuilder();
+		$tagsQuery = $qb->selectDistinct(['t.*', 'mt.imap_message_id'])
+			->from($this->getTableName(), 't')
+			->join('t', 'mail_message_tags', 'mt', $qb->expr()->eq('t.id', 'mt.tag_id', IQueryBuilder::PARAM_INT))
+			->where(
+				$qb->expr()->in('mt.imap_message_id', $qb->createParameter('ids'), IQueryBuilder::PARAM_STR_ARRAY),
+				$qb->expr()->eq('t.user_id', $qb->createNamedParameter($userId, IQueryBuilder::PARAM_STR)),
+				$qb->expr()->eq('t.imap_label', $qb->createNamedParameter($imapLabel, IQueryBuilder::PARAM_STR))
+			);
+
+		$messageIds = [];
+		foreach (array_chunk($ids, 1000) as $chunk) {
```

Comment:
nit: we could define a constant somewhere for the oracle chunking size to make this less of a magical number. It will help future devs understand the reason for the chunking more easily

Reply:
I'll open a new ticket for it

Reply:
Here: #5214